tests/foreman/destructive/test_installer.py

```diff
@@ -114,3 +126,30 @@ def test_installer_inventory_plugin_update(target_sat):
         )
     )
     assert '(current: true)' in verify_proxy_plugin_flag.stdout
+
+
+def test_positive_mismatched_satellite_fqdn(target_sat, set_wrong_fqdn):
+    """The satellite-installer should display the mismatched FQDN
+
+    :id: 264910ca-23c8-4192-a993-24bc04994a4c
+
+    :steps:
+        1. Set incorrect satellite hostname.
+        2. assert that output of 'facter fqdn' and 'hostname --fqdn' is different.
+        3. Run satellite-installer.
+        4. Assert that satellite-installer command displays mismatched FQDN.
+
+    :expectedresults: The satellite-installer should display the mismatched FQDN.
+
+    :CaseAutomation: Automated
+    """
+    facter_command_output = target_sat.execute('facter fqdn').stdout.strip()
+    hostname_command_output = target_sat.execute('hostname --fqdn').stdout.strip()
```

Comment:
I think the trick here is, using a shortname rather than an FQDN to generate the difference in the two:

```
[root@sat-6-10-qa-rhel7 vagrant]# hostname test
[root@sat-6-10-qa-rhel7 vagrant]# hostname -f
test
[root@sat-6-10-qa-rhel7 vagrant]# facter fqdn
test.wareagle.example.com
```

Reply:
Oh sorry, I see the `gen_string` is generating a short hostname. Seems then facter is unable to resolve the domain and punting back to just reporting the host. The logic facter uses is linked below, and the tricky part is figuring out which of those code paths is not triggering such that `facter fqdn` should be resolving the domain and thus getting a different output than `hostname`.


@ekohl @evgeni 

https://github.com/puppetlabs/facter/blob/main/lib/facter/resolvers/linux/hostname.rb

Reply:
I am thinking that this may be related to what is in `/etc/resolv.conf`, my local one:

```
# cat /etc/resolv.conf 
# Generated by NetworkManager
search wareagle.example.com
nameserver 192.168.121.1
```

I think that `search` line is what is allowing facter to report back a full FQDN rather than just the short hostname.

Reply:
Updated the test, waiting for the automation result.

Reply:
Because I was interested in how it really worked I also had to look at manpages. `man 1 hostname` states for "THE FQDN" bit:
> Technically: The FQDN is the name getaddrinfo(3) returns for the host name returned by gethostname(2).  The DNS domain name is the part after the first dot.

And then making some selection, though the manpage doesn't specify how it does this. It does say:

> If a machine has multiple network interfaces/addresses or is used in a mobile environment, then it may either have multiple FQDNs/domain names or none at all. Therefore avoid using hostname --fqdn, hostname --domain and dnsdomainname.  hostname --ip-address is subject to the same limitations so it should be avoided as well.

The shortened C version is (removing all error handling and other cases):
```c
struct addrinfo *res;
struct addrinfo hints;
char *p;

memset(&hints, 0, sizeof(struct addrinfo));
hints.ai_socktype = SOCK_DGRAM;
hints.ai_flags = AI_CANONNAME;

p = localhost(); # a wrapper for gethostname() with a safe buffer and cache

getaddrinfo(p, NULL, &hints, &res);

printf("%s\n", res->ai_canonname);
```

See https://sources.debian.org/src/hostname/3.23%2Bnmu1/hostname.c/ for the full source.  At least Fedora 36 uses `hostname` from Debian.

Reading `man 3 getaddrinfo` it states:

> getaddrinfo() returns one or more addrinfo structures

And `addrinfo` contains an entry `ai_next` so I think it's a linked list. That suggests uses the first entry that's returned.

I think the Python equivalent is:
```python
import socket
addrs = socket.getaddrinfo(socket.gethostname(), None, type=socket.SOCK_DGRAM, flags=socket.AI_CANONNAME)
print(addrs[0][3])
```